In [46]:
import pandas as pd

In [47]:
df = pd.read_csv("listings.csv")

In [48]:
df.isnull().sum()[df.isnull().sum() > 0]


description                        9
neighborhood_overview            670
host_name                         39
host_since                        39
host_location                    270
host_about                       509
host_response_time               137
host_response_rate               137
host_acceptance_rate             121
host_is_superhost                136
host_thumbnail_url                39
host_picture_url                  39
host_neighbourhood                76
host_listings_count               39
host_total_listings_count         39
host_verifications                39
host_has_profile_pic              39
host_identity_verified            39
neighbourhood                    670
neighbourhood_group_cleansed    1375
bathrooms                        238
bathrooms_text                     1
bedrooms                          76
beds                             240
price                            245
minimum_minimum_nights             1
maximum_minimum_nights             1
m

In [49]:
valores_nulos=df.isnull().sum()
valores_nulos

id                                                0
listing_url                                       0
scrape_id                                         0
last_scraped                                      0
source                                            0
                                               ... 
calculated_host_listings_count                    0
calculated_host_listings_count_entire_homes       0
calculated_host_listings_count_private_rooms      0
calculated_host_listings_count_shared_rooms       0
reviews_per_month                               336
Length: 79, dtype: int64

In [50]:
df2=df.copy()

Rellenamos con la media porque representa bien la tendencia general de reseñas mensuales

In [51]:
df2["reviews_per_month"]=df2["reviews_per_month"].fillna(round(df2["reviews_per_month"].mean(),1))

In [52]:
valores_nulos=df2.isnull().sum()
valores_nulos

id                                              0
listing_url                                     0
scrape_id                                       0
last_scraped                                    0
source                                          0
                                               ..
calculated_host_listings_count                  0
calculated_host_listings_count_entire_homes     0
calculated_host_listings_count_private_rooms    0
calculated_host_listings_count_shared_rooms     0
reviews_per_month                               0
Length: 79, dtype: int64

Rellenamos con "N/A" porque es texto y no tiene sentido hacer promedio y no nos podemos inverntar una descripción

In [53]:
df2["description"] =df2["description"].fillna("N/A")

Se relaiza lo mismo, igual rellenamos con "N/A" por ser texto descriptivo N/A es utilizado ya que el mismo excel lo utiliza

In [54]:
df2["neighborhood_overview"] =df2["neighborhood_overview"].fillna("N/A")


In [55]:
#df2["host_name"] =df2["host_name"].fillna("N/A")


In [56]:
#df2["host_since"] =df2["host_since"].fillna(method="bfill") 


host_response_time, host_response_rate, host_acceptance_rate: rellenamos con "N/A" ya que son datos de texto

In [57]:
df2["host_location"] =df2["host_location"].fillna("N/A")


In [58]:
df2["host_about"] =df2["host_about"].fillna("N/A")


In [59]:
df2["host_response_time"] =df2["host_response_time"].fillna("N/A")

In [60]:
df2["host_response_rate"] =df2["host_response_rate"].fillna("N/A")

In [61]:
df2["host_acceptance_rate"] =df2["host_acceptance_rate"].fillna("N/A")

Utilicé backward fill porque es binario y suele repetirse por anfitrión

In [62]:
df2["host_is_superhost"] =df2["host_is_superhost"].fillna(method="bfill") 

/var/folders/rn/3d02syds5p99hnhh5p4p92jh0000gn/T/ipykernel_26839/1239376417.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df2["host_is_superhost"] =df2["host_is_superhost"].fillna(method="bfill")


Se rellena con "N/A", información textual

In [63]:
df2["host_neighbourhood"] =df2["host_neighbourhood"].fillna("N/A")


Rellenamos con la mediana para no afectar por valores extremos

In [64]:
df2["host_listings_count"]=df2["host_listings_count"].fillna(round(df["host_listings_count"].median(),1))


In [65]:
df2["host_total_listings_count"]=df2["host_total_listings_count"].fillna(round(df["host_total_listings_count"].median(),1))

Se rellena con "None" porque es un listado de verificaciones

In [66]:
df2["host_verifications"] =df2["host_verifications"].fillna("None")

Se rellena con "N/A", texto descriptivo de zona

In [67]:
df2["neighbourhood"] =df2["neighbourhood"].fillna("N/A")


Se rellenan con 0, ya que representa cantidad y el excel ya cuenta con este resultado

In [68]:
df2["bathrooms"]= df2["bathrooms"].fillna(0) 

Se rellena con "0 shared baths" para mantener consistencia textual y si no hay datos cuenta como 0

In [69]:
df2["bathrooms_text"] =df2["bathrooms_text"].fillna("0 shared baths")

Se rellena con la media porque son cantidades y queremos mantener la tendencia general

In [70]:
df2["bedrooms"]=df2["bedrooms"].fillna(round(df2["bedrooms"].mean(),1))


In [71]:
df2["beds"]=df2["beds"].fillna(round(df2["beds"].mean(),1))


primero quitamos signos de $ y convertimos a float, luego rellenamos con la media

In [72]:
df2["price"] = df2["price"].replace('[\$,]', '', regex=True)
df2["price"] = df2["price"].astype(float)
df2

<>:1: SyntaxWarning: invalid escape sequence '\$'
<>:1: SyntaxWarning: invalid escape sequence '\$'
/var/folders/rn/3d02syds5p99hnhh5p4p92jh0000gn/T/ipykernel_26839/3033783479.py:1: SyntaxWarning: invalid escape sequence '\$'
  df2["price"] = df2["price"].replace('[\$,]', '', regex=True)


id                                       listing_url  \
0                    8521                 https://www.airbnb.com/rooms/8521   
1                   11169                https://www.airbnb.com/rooms/11169   
2                   19581                https://www.airbnb.com/rooms/19581   
3                   27498                https://www.airbnb.com/rooms/27498   
4                   79762                https://www.airbnb.com/rooms/79762   
...                   ...                                               ...   
1370  1442190452341432060  https://www.airbnb.com/rooms/1442190452341432060   
1371  1443555192209357989  https://www.airbnb.com/rooms/1443555192209357989   
1372  1443585853586366833  https://www.airbnb.com/rooms/1443585853586366833   
1373  1447874951198409920  https://www.airbnb.com/rooms/1447874951198409920   
1374  1447875564779712913  https://www.airbnb.com/rooms/1447875564779712913   

           scrape_id last_scraped       source  \
0     20250625193454   2025-06-26  city scrape   
1     20250625193454   2025-06-27  city scrape   
2     20250625193454   2025-06-26  city scrape   
3     20250625193454   2025-06-27  city scrape   
4     20250625193454   2025-06-26  city scrape   
...              ...          ...          ...   
1370  20250625193454   2025-06-26  city scrape   
1371  20250625193454   2025-06-26  city scrape   
1372  20250625193454   2025-06-26  city scrape   
1373  20250625193454   2025-06-26  city scrape   
1374  20250625193454   2025-06-26  city scrape   

                                                  name  \
0     SunsplashedSerenity walk to Harvard & Fresh Pond   
1        Lovely Studio Room: Available for long w/ends   
2                             Furnished suite, Windsor   
3                      Furnished suite 2 @ the Windsor   
4                    Cambridge Getaway @ Harvard & MIT   
...                                                ...   
1370         Cozy 2-Bed Apt Near MIT/Harvard + Parking   
1371                           Walk to Harvard & MIT Y   
1372                          Walk to Harvard & MIT XX   
1373                      Modern Apt In Kendall Square   
1374                               Modern Apt with Gym   

                                            description  \
0     An elegant, sun-splashed, 2 bedroom (+2offices...   
1     Large sunny room w kitchenette & bath. Foam ma...   
2     Welcome to Area IV! We are located, convenient...   
3     Welcome to Area IV! We are located, convenient...   
4     Charming 2-bedroom apartment on the third floo...   
...                                                 ...   
1370  Our 2bd, 1ba is a 5 minute walk to Central Squ...   
1371  PRIVATE ROOM <br />Clean and comfortable room ...   
1372  PRIVATE ROOM <br />Clean and comfortable room ...   
1373  Welcome to Proto Kendall Square Apartments, wh...   
1374  Welcome to Proto Kendall Square Apartments, wh...   

                                  neighborhood_overview  \
0     Huron Village is known for its charm.  We have...   
1     The neighborhood is quiet and friendly and our...   
2                                                   N/A   
3                                                   N/A   
4     Annmarie and I have lived in this area for ove...   
...                                                 ...   
1370  Located in the heart of central square. This i...   
1371                                                N/A   
1372                                                N/A   
1373  Kendall Square in Cambridge, Massachusetts, of...   
1374  Kendall Square in Cambridge, Massachusetts, of...   

                                            picture_url   host_id  ...  \
0     https://a0.muscache.com/pictures/30536/072e0a5...    306681  ...   
1     https://a0.muscache.com/pictures/miso/Hosting-...     40965  ...   
2     https://a0.muscache.com/pictures/188f1b4b-f37b...     74249  ...   
3     https://a0.muscache.com/pictures/bab30c3c-ff3c...     74

In [73]:
df2["price"]=df2["price"].fillna(round(df2["price"].mean(),1))

Rellenamos con "t" como valor por defecto ya que el excel solo tiene esa opción

In [74]:
df2["has_availability"] =df2["has_availability"].fillna("t")


Se rellena con mediana para no afectar por valores extremos

In [75]:
df2["estimated_revenue_l365d"]=df2["estimated_revenue_l365d"].fillna(round(df["estimated_revenue_l365d"].median(),1))


Se utiliza el backward y forward fill para mantener continuidad temporal

In [76]:
df2["first_review"] =df2["first_review"].fillna(method="bfill") 

/var/folders/rn/3d02syds5p99hnhh5p4p92jh0000gn/T/ipykernel_26839/226337311.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df2["first_review"] =df2["first_review"].fillna(method="bfill")


In [77]:
df2["first_review"] =df2["first_review"].fillna(method="ffill") 

/var/folders/rn/3d02syds5p99hnhh5p4p92jh0000gn/T/ipykernel_26839/1983289630.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df2["first_review"] =df2["first_review"].fillna(method="ffill")


In [78]:
df2["last_review"] =df2["last_review"].fillna(method="bfill") 

/var/folders/rn/3d02syds5p99hnhh5p4p92jh0000gn/T/ipykernel_26839/3395681466.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df2["last_review"] =df2["last_review"].fillna(method="bfill")


In [79]:
df2["last_review"] =df2["last_review"].fillna(method="ffill") 

/var/folders/rn/3d02syds5p99hnhh5p4p92jh0000gn/T/ipykernel_26839/3777977187.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df2["last_review"] =df2["last_review"].fillna(method="ffill")


Se rellena con 0, ya que es numérico y no tiene información faltante real

In [80]:
df2["license"]= df2["license"].fillna(0) 

Estas columnas tienen todos nulos 1375 registros sin datos, por lo que no aportan variables

In [81]:
df2 = df2.drop(columns=["neighbourhood_group_cleansed", "calendar_updated"])

Solo hay 1 fila con nulos, por lo que eliminarla no afecta el dataset.

In [82]:
df2 = df2.dropna(subset=["minimum_minimum_nights", "maximum_minimum_nights", "minimum_maximum_nights", "maximum_maximum_nights"])


Rellenar los nulos con la media permite mantener el valor general de la columna sin sesgar los datos con valores extremos.

In [83]:
review_cols = ["review_scores_rating", "review_scores_accuracy", "review_scores_cleanliness", "review_scores_checkin", "review_scores_communication", "review_scores_location", "review_scores_value"]
for col in review_cols:
    df2[col] = df2[col].fillna(round(df2[col].mean(),1))


Solo hay 39 filas con nulos, lo que representa una pequeña parte del dataset al igual que no se puede rellenar de manera confiable (no tiene sentido inventar nombres, fechas o fotos).

In [84]:
host_cols = ["host_name", "host_since", "host_thumbnail_url", "host_picture_url", "host_has_profile_pic", "host_identity_verified"]
df2 = df2.dropna(subset=host_cols)


In [85]:
df2.isnull().sum()[df2.isnull().sum() > 0]


Series([], dtype: int64)

In [86]:
df2.to_csv("listings_limpio.csv")